test/api_bridge_test.rb

```diff
@@ -0,0 +1,165 @@
+require 'test_helper'
+
+class Orchid::Test < ActiveSupport::TestCase
+  # Note: order of test assert_equal is "expected" and "actual"
+
+  def before_setup
+    super
+    @api = ApiBridge::Query.new("https://test.unl.edu/v1",
+      ["creator.name", "date.year", "subcategory"],
+      { "num" => 5, "sort" => "title|asc" }
+    )
+  end
+
+  # INITIALIZATION
+
+  test "initialization: bad url" do
+    assert_raises do
+      ApiBridge::Query.new("bad_url")
+    end
+  end
+
+  test "initialization: no options" do
+    api = ApiBridge::Query.new("https://validurl.unl.edu/v1")
+    assert api
+    assert_equal [], api.app_facets
+    assert_equal({ "num" => 50, "facet" => [] }, api.app_options)
+  end
+
+  test "initialization: override options" do
+    assert @api
+    facets = ["creator.name", "date.year", "subcategory"]
+    assert_equal facets, @api.app_facets
+    assert_equal(
+      { "num" => 5, "sort" => "title|asc", "facet" => facets },
+      @api.app_options
+    )
+  end
+
+  test "calc_start" do
+    # setting pages and nums
+    opts = @api.calc_start({"page" => "2", "num" => "3"})
+    assert_equal 3, opts["start"]
+    opts = @api.calc_start({"page" => "10", "num" => "50"})
+    assert_equal 450, opts["start"]
+    # no page set
+    opts = @api.calc_start({"num" => "20"})
+    assert_equal 0, opts["start"]
+    # illegal page
+    opts = @api.calc_start({"page" => "-1", "num" => "10"})
+    assert_equal 0, opts["start"]
+  end
+
+  test "create_item_url" do
+    url = @api.create_item_url("cat.let1728")
+    assert_equal "https://test.unl.edu/v1/item/cat.let1728", url
+  end
+
+  test "create_items_url" do
+    # just a facet
+    url = @api.create_items_url({ "facet" => ["date.year", "source"] })
+    assert_equal "https://test.unl.edu/v1/items?facet[]=date.year&facet[]=source", url
+    # a filter that needs some encoding
+    url = @api.create_items_url({ "f" => ["creator.name|\"Yosemite Sam\": Worst Name Ever"]})
```

Comment:
The rootin'est, tootin'est test there ever was! :cowboy_hat_face:

Reply:
🐴 🤠